### PR TITLE
Handle If-None-Match and If-Match for all endpoint and pass through the corresponding header to Sync

### PR DESCRIPTION
**Pull request by Natim**

### CODE REVIEW

**Comment by Natim** [comment]
@michielbdejong does this fix your cache issues with Kinto.js?

**Comment by michielbdejong** [comment]
I just tried this branch and got a 500 error:

```
serving on http://0.0.0.0:8000

ERROR:venusian:"GET   /v1/buckets/syncto/collections/meta/records" ? (? ms) get_records() got an unexpected keyword argument 'headers' lang=en-US,en;q=0.5; exception=Traceback (most recent call last):
  File "/Users/Michiel/syncto/.venv/lib/python2.7/site-packages/pyramid/tweens.py", line 21, in excview_tween
    response = handler(request)
  File "/Users/Michiel/syncto/.venv/lib/python2.7/site-packages/pyramid/router.py", line 163, in handle_request
    response = view_callable(context, request)
  File "/Users/Michiel/syncto/.venv/lib/python2.7/site-packages/pyramid/config/views.py", line 596, in __call__
    return view(context, request)
  File "/Users/Michiel/syncto/.venv/lib/python2.7/site-packages/pyramid/config/views.py", line 329, in attr_view
    return view(context, request)
  File "/Users/Michiel/syncto/.venv/lib/python2.7/site-packages/pyramid/config/views.py", line 305, in predicate_wrapper
    return view(context, request)
  File "/Users/Michiel/syncto/.venv/lib/python2.7/site-packages/pyramid/config/views.py", line 355, in rendered_view
    result = view(context, request)
  File "/Users/Michiel/syncto/.venv/lib/python2.7/site-packages/pyramid/config/views.py", line 501, in _requestonly_view
    response = view(request)
  File "/Users/Michiel/syncto/.venv/lib/python2.7/site-packages/cornice/service.py", line 569, in wrapper
    response = view_(request)
  File "/Users/Michiel/syncto/syncto/views/collection.py", line 62, in collection_get
    headers=headers, **params)
TypeError: get_records() got an unexpected keyword argument 'headers'; uid=None; errno=None; agent=Mozilla/5.0 (Mobile; rv:42.0) Gecko/20100101 Firefox/42.0; authn_type=None
```

I do not get this error in the master branch.

**Comment by Natim** [comment]
Yes you need to update the syncclient to its last release.

```
.venv/bin/pip install -Ue .
```

File: syncto/headers.py [hunk 1]
```diff
@@ -1,11 +1,31 @@
-def convert_headers(sync_raw_response, syncto_response):
+def import_headers(syncto_request, sync_request_headers=None):
+    """Convert incoming Kinto headers into Sync headers."""
+    request_headers = syncto_request.headers
+    headers = sync_request_headers or {}
+
+    if 'If-Match' in request_headers:
+        headers['X-If-Unmodified-Since'] = '%.2f' % (
+            int(request_headers['If-Match']) / 1000.0)
```
Comment (leplatrem) [inline syncto/headers.py]: `If-Match` is not supposed to contain quotes ? `int('"12"')` won't work, do I miss something ?

File: syncto/tests/test_functional.py [hunk 2]
```diff
@@ -140,7 +140,30 @@ def test_collection_handle_since_parameter(self):
                              '_sort': 'newest'},
                      headers=self.headers, status=200)
         self.sync_client.return_value.get_records.assert_called_with(
-            "tabs", full=True, newer='14377478425.70', sort='newest')
+            "tabs", full=True, newer='14377478425.70', sort='newest',
+            headers={})
+
+    def test_collection_handle_if_none_match_headers(self):
+        headers = self.headers.copy()
+        headers['If-None-Match'] = '14377478425700'
```
Comment (leplatrem) [inline syncto/tests/test_functional.py]: Yes, here you need double quotes

File: syncto/tests/test_functional.py [hunk 3]
```diff
@@ -212,6 +236,17 @@ def test_can_delete_record(self):
         self.sync_client.return_value.delete_record.return_value = None
         self.app.delete(RECORD_URL, headers=self.headers, status=204)
 
+    def test_can_delete_record_handle_headers(self):
+        self.sync_client.return_value.delete_record.return_value = None
+
+        headers = self.headers.copy()
+        headers['If-Match'] = '14377478425700'
```
Comment (leplatrem) [inline syncto/tests/test_functional.py]: ditto

File: syncto/tests/test_functional.py [hunk 4]
```diff
@@ -140,7 +140,30 @@ def test_collection_handle_since_parameter(self):
...
+                     params={'_since': '14377478425700',
+                             '_sort': 'newest'},
+                     headers=headers, status=200)
+        self.sync_client.return_value.get_records.assert_called_with(
+            "tabs", full=True, newer='14377478425.70', sort='newest',
+            headers={'X-If-Modified-Since': '14377478425.70'})
+
+    def test_collection_handle_if_match_headers(self):
+        headers = self.headers.copy()
+        headers['If-Match'] = '14377478425700'
```
Comment (leplatrem) [inline syncto/tests/test_functional.py]: ditto

File: syncto/tests/test_functional.py [hunk 5]
```diff
@@ -140,7 +140,30 @@ def test_collection_handle_since_parameter(self):
                              '_sort': 'newest'},
                      headers=self.headers, status=200)
         self.sync_client.return_value.get_records.assert_called_with(
-            "tabs", full=True, newer='14377478425.70', sort='newest')
+            "tabs", full=True, newer='14377478425.70', sort='newest',
+            headers={})
+
+    def test_collection_handle_if_none_match_headers(self):
```
Comment (leplatrem) [inline syncto/tests/test_functional.py]: nit: handle**s**

File: syncto/tests/test_functional.py [hunk 6]
```diff
@@ -212,6 +236,17 @@ def test_can_delete_record(self):
         self.sync_client.return_value.delete_record.return_value = None
         self.app.delete(RECORD_URL, headers=self.headers, status=204)
 
+    def test_can_delete_record_handle_headers(self):
```
Comment (leplatrem) [inline syncto/tests/test_functional.py]: nit: test «delete handles if match header»

File: syncto/tests/test_functional.py [hunk 7]
```diff
@@ -244,6 +279,17 @@ def test_can_put_valid_record(self):
         self.app.put_json(RECORD_URL, RECORD_EXAMPLE,
                           headers=self.headers, status=200)
 
+    def test_put_record_handle_headers(self):
```
Comment (leplatrem) [inline syncto/tests/test_functional.py]: ditto: handles if none match header

File: syncto/tests/test_functional.py [hunk 8]
```diff
@@ -260,10 +306,17 @@ def test_put_return_a_503_in_case_of_unknown_error(self):
     def test_put_return_a_400_in_case_of_bad_request(self):
```
Comment (leplatrem) [inline syncto/tests/test_functional.py]: bad request returns alert and backoff response headers

**Comment by michielbdejong** [comment]
Ah, thanks! Now, syncing once works. :) But when syncing a second time, I still get a 500:

```
ERROR:venusian:"GET   /v1/buckets/syncto/collections/meta/records?_since=1439218398550" ? (? ms) invalid literal for int() with base 10: '"1439218398550"' lang=en-US,en;q=0.5; exception=Traceback (most recent call last):
  File "/Users/Michiel/syncto/.venv/lib/python2.7/site-packages/pyramid/tweens.py", line 21, in excview_tween
    response = handler(request)
  File "/Users/Michiel/syncto/.venv/lib/python2.7/site-packages/pyramid/router.py", line 163, in handle_request
    response = view_callable(context, request)
  File "/Users/Michiel/syncto/.venv/lib/python2.7/site-packages/pyramid/config/views.py", line 596, in __call__
    return view(context, request)
  File "/Users/Michiel/syncto/.venv/lib/python2.7/site-packages/pyramid/config/views.py", line 329, in attr_view
    return view(context, request)
  File "/Users/Michiel/syncto/.venv/lib/python2.7/site-packages/pyramid/config/views.py", line 305, in predicate_wrapper
    return view(context, request)
  File "/Users/Michiel/syncto/.venv/lib/python2.7/site-packages/pyramid/config/views.py", line 355, in rendered_view
    result = view(context, request)
  File "/Users/Michiel/syncto/.venv/lib/python2.7/site-packages/pyramid/config/views.py", line 501, in _requestonly_view
    response = view(request)
  File "/Users/Michiel/syncto/.venv/lib/python2.7/site-packages/cornice/service.py", line 569, in wrapper
    response = view_(request)
  File "/Users/Michiel/syncto/syncto/views/collection.py", line 23, in collection_get
    headers = import_headers(request)
  File "/Users/Michiel/syncto/syncto/headers.py", line 15, in import_headers
    int(request_headers['If-None-Match']) / 1000.0)
ValueError: invalid literal for int() with base 10: '"1439218398550"'; uid=None; errno=None; agent=Mozilla/5.0 (Mobile; rv:42.0) Gecko/20100101 Firefox/42.0; authn_type=None
```

**Comment by Natim** [comment]
Yes this is related to @leplatrem review. I am going to fix that thanks for testing @michielbdejong

**Comment by Natim** [comment]
I took your remarks into account, let me also add tests for wrong If-None-Match and If-Match values.

**Comment by michielbdejong** [comment]
The error now changed to:

```
ERROR:venusian:"GET   /v1/buckets/syncto/collections/meta/records?_since=1439218398550" ? (? ms) 
Expected object or value lang=en-US,en;q=0.5; exception=Traceback (most recent call last):
  File "/Users/Michiel/syncto/.venv/lib/python2.7/site-packages/pyramid/tweens.py", line 21, in 
excview_tween
[...]
```

This is the request that causes it:

```
curl 'http://localhost:8000/v1/buckets/syncto/collections/meta/records?_since=1439218398550' \
-H 'Host: localhost:8000' \
-H 'User-Agent: Mozilla/5.0 (Mobile; rv:42.0) Gecko/20100101 Firefox/42.0' \
-H 'Accept: application/json' \
-H 'Accept-Language: en-US,en;q=0.5' \
--compressed \
-H 'If-None-Match: "1439218398550"' \
-H 'x-client-state: 518fef27c6bbc0220aab0f00b1a37308' \
-H 'Content-Type: application/json' \
-H 'Authorization: BrowserID REMOVED==' \
-H 'origin: app://sync.gaiamobile.org' \
-H 'Connection: keep-alive'
```

(apparently it's choking on the `Accept-Language` header?)

**Comment by Natim** [comment]
Ok this is a problem in `syncclient` try to read an empty body.

```
 File "syncclient/client.py", line 112, in _request
    return self.raw_resp.json()
  File "requests/models.py", line 819, in json
    return json.loads(self.text, **kwargs)
ValueError: Expected object or value; uid=None; errno=None; agent=HTTPie/0.9.2; authn_type=None
```

**Comment by Natim** [comment]
See https://github.com/mozilla-services/syncclient/pull/14

**Comment by michielbdejong** [comment]
It's working now, I'm seeing 304's when I test it. :) Just one thing: The server log reports 304s as Client Errors, even though they are not really errors, they will occur during normal operation. So maybe you can avoid that, otherwise the server logs would fill up with (thousands of) messages like:

```
HTTPError: 304 Client Error: Not Modified for url: https://sync-237-us-west-2.sync.services.mozilla.com/1.5/26629823/storage/meta?full=True&newer=1439218398.55; uid=None; errno=None; agent=Mozilla/5.0 (Mobile; rv:44.0) Gecko/20100101 Firefox/44.0; authn_type=None
ERROR:venusian:"GET   /v1/buckets/syncto/collections/crypto/records?_since=1439218393690" ? (? ms) 304 Client Error: Not Modified for url: https://sync-237-us-west-2.sync.services.mozilla.com/1.5/26629823/storage/crypto?full=True&newer=1439218393.69 lang=en-US,en;q=0.5; exception=Traceback (most recent call last):
  File "/Users/Michiel/syncto/.venv/lib/python2.7/site-packages/pyramid/tweens.py", line 21, in excview_tween
    response = handler(request)
...
```

**Comment by Natim** [comment]
Oh nice catch.
